### PR TITLE
Migrate Sell > Catalog > Discounts > Catalog Price Rules list

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Migrate Sell > Catalog > Discounts > Catalog Price Rules list without actions. Add/Edit Actions are implemented in #13716 
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10549
| How to test?  | Use `admin-dev/index.php/sell/catalog/catalog-price-rules` to see the working list. Behavior should stay the same as in legacy page. Except add/edit actions are implemented in another PR.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/13584)

### CODE REVIEW

File: src/Core/Domain/CatalogPriceRule/Command/BulkDeleteCatalogPriceRuleCommand.php
Comment (sarjon) [inline src/Core/Domain/CatalogPriceRule/Command/BulkDeleteCatalogPriceRuleCommand.php:27]: @matks wdyt about `Domain\CatalogPriceRule`? In object model it's called `SpecificPriceRule`, but in BO it's "Catalog Price Rule" and it can be found under "Discounts" section.

Besides that, there is `CartRule` as well that fits "Discount" description.

File: src/Core/Domain/CatalogPriceRule/Exception/CatalogPriceRuleException.php
Comment (sarjon) [inline src/Core/Domain/CatalogPriceRule/Exception/CatalogPriceRuleException.php]: Shouldn't it be `subdomain` instead?

File: src/Core/Grid/Definition/Factory/CatalogPriceRuleGridDefinitionFactory.php
Comment (sarjon) [inline src/Core/Grid/Definition/Factory/CatalogPriceRuleGridDefinitionFactory.php]: Is `$this->queryBus` used?

File: src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/_catalog.yml
Comment (sarjon) [inline src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/_catalog.yml]: Why not `catalog-price-rules`?

File: src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/catalog_price_rule.yml
Comment (sarjon) [inline src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/catalog_price_rule.yml:8]: `_legacy_link` is missing.

File: src/PrestaShopBundle/Resources/config/services/adapter/catalog_price_rule.yml
Comment (sarjon) [inline src/PrestaShopBundle/Resources/config/services/adapter/catalog_price_rule.yml]: You can split into 2 lines, so it's more readable.

File: src/PrestaShopBundle/Resources/config/services/core/grid/grid_data_factory.yml
Comment (sarjon) [inline src/PrestaShopBundle/Resources/config/services/core/grid/grid_data_factory.yml]: Too many lines I think.

**Comment by zuk3975** [comment]
@matks, ready for review

File: src/Core/Grid/Definition/Factory/CatalogPriceRuleGridDefinitionFactory.php
Comment (zuk3975) [inline src/Core/Grid/Definition/Factory/CatalogPriceRuleGridDefinitionFactory.php]: @matks, for this specific page, exists such timestamps '0000-00-00 00:00:00'.
DateTimeColumn renders such time stamp in to strange numbers  (image bellow). Any offers for that?
Screenshot from 2019-05-15 15-37-59.

File: src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php
Comment (zuk3975) [inline src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php:179]: @matks, according to legacy- reduction field in list is shown rounded to decimals, so filtering float which is, for example 50.0005 wont find it as exact match, but user will see 50.00 in unfiltered list. Seems like UX issue to me. 
Currently i just didn't round the numbers. Any other offers?

File: src/Core/Domain/CatalogPriceRule/Command/BulkDeleteCatalogPriceRuleCommand.php
Comment (matks) [reply src/Core/Domain/CatalogPriceRule/Command/BulkDeleteCatalogPriceRuleCommand.php:27]: Legacy code always consistent in naming 😄 

`CatalogPriceRule` is better than `SpecificPriceRule` . `SpecificPriceRule` is very unclear.
Wondering about the `Discount` domain 🤔 

@eternoendless ?

File: src/Core/Grid/Definition/Factory/CatalogPriceRuleGridDefinitionFactory.php
Comment (matks) [reply src/Core/Grid/Definition/Factory/CatalogPriceRuleGridDefinitionFactory.php]: Usually, when `0000-00-00 00:00:00` value is used in a database, it's an alias for `null` but the database does not allow nullable values.

So I suggest to display it like a blank value. Nothing.

@colinegin @MatShir Do you agree ? Display a blank space instead of this irrelevant `-0001-11-30 00:00` ?

@zuk3975 Create a php constant `const DATETIME_NULL_VALUE = '0000-00-00 00:00:00';` to handle it so you can do `if ($date === self::DATETIME_NULL_VALUE)` in the code

File: src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php
Comment (matks) [inline src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php:154]: I'm not sure but this `Connection::PARAM_INT_ARRAY` might not work with all SQL implementations 🤔 can you check ? We target mysql 5.7 to mysql 8

It's not clear but I remember having issues with this. Maybe with sqlite though (in this case, we dont care).

File: src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php
Comment (matks) [reply src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php:179]: @eternoendless This is a Grid migration PR. Grid do not depend on legacy code, only on the legacy SQL schema.

I think we can introduce https://github.com/PrestaShop/decimal here. Do you agree ?

Comment (eternoendless) [reply src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php:179]: Use Decimal as much as you can, specially when performing price operations.

Comment (eternoendless) [reply src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php:179]: Also, when rounding, you must keep in mind the rounding mode as configured in the shop settings.

File: src/Core/Domain/CatalogPriceRule/Command/BulkDeleteCatalogPriceRuleCommand.php
Comment (eternoendless) [inline src/Core/Domain/CatalogPriceRule/Command/BulkDeleteCatalogPriceRuleCommand.php]: You can sign this parameter as an array

File: src/Core/Domain/CatalogPriceRule/Command/BulkDeleteCatalogPriceRuleCommand.php
Comment (eternoendless) [reply src/Core/Domain/CatalogPriceRule/Command/BulkDeleteCatalogPriceRuleCommand.php:27]: `CatalogPriceRule` sounds right regarding the end-user terminology.

File: src/Core/Grid/Definition/Factory/CatalogPriceRuleGridDefinitionFactory.php
Comment (zuk3975) [reply src/Core/Grid/Definition/Factory/CatalogPriceRuleGridDefinitionFactory.php]: https://github.com/PrestaShop/PrestaShop/pull/13584/commits/b87457615bb71c93d4368028050b6beb0abcbd07 wdyt?

File: src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php
Comment (zuk3975) [reply src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php:179]: Confused with the rounding :confused:. According to https://github.com/PrestaShop/PrestaShop/issues/12223 I CLDR should handle the prices rounding, but could it be also used for reduction :thinking: ?. (if possible maybe it's worth introducing some kind of DecimalColumn/PriceColumn ? also the rounding should also be considered when filtering).

File: src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php
Comment (zuk3975) [reply src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php:154]: Our current Doctrine/DBAL version declares that it supports mysql 5.0 and above, so i guess it's not a problem

File: src/Core/Util/DateTime/DateTime.php
Comment (matks) [inline src/Core/Util/DateTime/DateTime.php]: I'd say:
- either we dont care that this class can be initialized, because it's useless so if a module/agency developer uses it, I dont see what usage he can do about it 😄 
- if we dont want people to be mistaken between `\DateTime` and `PrestaShop\PrestaShop\Core\Util\DateTime\DateTime` then we can even further throw a RuntimeException in the __construct() with the error message

File: src/Core/Grid/Definition/Factory/CatalogPriceRuleGridDefinitionFactory.php
Comment (matks) [reply src/Core/Grid/Definition/Factory/CatalogPriceRuleGridDefinitionFactory.php]: See answer https://github.com/PrestaShop/PrestaShop/pull/13584/files#r301597349 😉

File: src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/catalog_price_rule.yml
Comment (matks) [inline src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/catalog_price_rule.yml:19]: Let's not forget these `@todo` 😄

File: admin-dev/themes/new-theme/js/components/grid/extension/filters-submit-button-enabler-extension.js
Comment (matks) [inline admin-dev/themes/new-theme/js/components/grid/extension/filters-submit-button-enabler-extension.js]: No breaking change possible for the pages that already use this extension ?

Comment (zuk3975) [reply admin-dev/themes/new-theme/js/components/grid/extension/filters-submit-button-enabler-extension.js]: Noticed that the selector of datepicker was incorrect :smile:  and in fact its not even needed (as it's already selected as input). Anyway, this only adds additional datepicker event and it is the first migrated page to have datepicker filter in grid so i doubt it could cause bc break.

File: src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/catalog_price_rule.yml
Comment (zuk3975) [reply src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/catalog_price_rule.yml:19]: It must be gone after this PR is rebased with create/edit PR for sure :smile:

File: src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php
Comment (matks) [reply src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php:179]: Sorry @zuk3975 my answer is irrelevant
I think this is an issue that must be discussed further with UX and Product, in another PR. I added it to the todo-list https://github.com/PrestaShop/PrestaShop/issues/10548 so we wont forget

**Review by matks — DISMISSED** [review]

**Comment by matks** [comment]
@zuk3975 We can go for QA once you rebase 😄

**Comment by zuk3975** [comment]
@matks rebased. Travis is mad for some reason :smile:

**Comment by sarahdib** [comment]
Hello @zuk3975 

1. The add a catalog price rule button isn't working
2. The edit button isn't working
image
3. on my macbook pro screen the date field isn't display properly @TristanLDD what do you think ?
image
4. In multistore -> I select the first shop in another page (like product) then I go http://localhost/PrestaShop/develop/admin-dev/index.php/sell/catalog/catalog-price-rules/?_token=cWywt5BTWVdX_1-mql5ka-nHHQraFZ_iN08kplZY6V4 in the list of catalog price rule the name of the second shop isn't display
image
But if I'm in the context all shop in another page then I go to http://localhost/PrestaShop/develop/admin-dev/index.php/sell/catalog/catalog-price-rules/?_token=cWywt5BTWVdX_1-mql5ka-nHHQraFZ_iN08kplZY6V4
in the list of catalog price rule the name of the second shop is display
image
5. the result of the search by date isn't correct : 
image
image

**Comment by zuk3975** [comment]
@sarahdib thanks and sorry for so many issues :sweat_smile: .
1 and 2 (actions) are not supposed to work in this PR. Those are done in #13716.
Also I've noticed that range filters should also have time pickers.
For others, i'll check.
**Edit**: issue 4 and 5 fixed.
Also added hours selection to date range filter, but its icons are missing. That might be related to issue #14748, so I think this PR needs to wait for it.

**Comment by sarahdib** [comment]
hello @zuk3975 
Don't worry :) it's my job to find issue :)  

You solved problem #4 & #5 solved ?

I'm waiting for @matks review :)

**Comment by zuk3975** [comment]
> hello @zuk3975
> Don't worry :) it's my job to find issue :)
> 
> You solved problem #4 & #5 solved ?
> 
> I'm waiting for @matks review :)

yes, 4 and 5 is solved. However I would like to put this pr aside until #14748 is fixed as updated in description.

**Comment by sarjon** [comment]
> However I would like to put this pr aside until #14748 is fixed as updated in description.

https://github.com/PrestaShop/PrestaShop/issues/14748 is fixed. Travis needs to be restarted and PR should be ready for QA.

**Comment by zuk3975** [comment]
> > However I would like to put this pr aside until #14748 is fixed as updated in description.
> 
> #14748 is fixed. Travis needs to be restarted and PR should be ready for QA.

Still need to wait until the fix reaches develop branch, then ill need to re-check if picker is working with hours format

**Comment by matks** [comment]
> > > However I would like to put this pr aside until #14748 is fixed as updated in description.
> > 
> > 
> > #14748 is fixed. Travis needs to be restarted and PR should be ready for QA.
> 
> Still need to wait until the fix reaches develop branch, then ill need to re-check if picker is working with hours format

It's done now 👍

**Comment by zuk3975** [comment]
@matks, rebased and adjusted datepicker to show hours. My job is done here :smile:

**Comment by matks** [comment]
> @matks, rebased and adjusted datepicker to show hours. My job is done here 😄

**Review by matks — DISMISSED** [review]

**Comment by zuk3975** [comment]
@matks, sorry dismissed review with one commit as i noticed forgotten commented-out code :sweat_smile: . Nothing else changed.

**Review by matks — DISMISSED** [review]

**Comment by matks** [comment]
> @matks, sorry dismissed review with one commit as i noticed forgotten commented-out code 😅 . Nothing else changed.

😞

**Comment by sarahdib** [comment]
Hello @zuk3975 
When I search with two filters there is no result found 
like
- When I search by name & ID there is no result
https://drive.google.com/file/d/1g9nbsEdkKFnO3t9QWU3Zo_qWfvThJS0Y/view
Same problem withe name & Shop
https://drive.google.com/file/d/1g9nbsEdkKFnO3t9QWU3Zo_qWfvThJS0Y/view

Result with legacy : https://drive.google.com/file/d/1BI0iSiEujb4cd6uDSgmDrdOZ_dHoLyFe/view


- When I try to order by groupe there is an error : 

https://drive.google.com/file/d/1k_Sm7x8HPiSxfT_Zt_3_WlZPlRZpxire/view
Same problem with order by date

**Comment by zuk3975** [comment]
@sarahdib, thank you.
There were some issues with filter names (apparently couple of them were reserved mysql keywords)...
Fixed. Sorting and filtering seems to work as expected now.

**Review by matks — DISMISSED** [review]

**Comment by matks** [comment]
> There were some issues with filter names (apparently couple of them were reserved mysql keywords)...

That old funny pal mysql 😄trolling us

**Comment by sarahdib** [comment]
@zuk3975 Thank you :)

Can you resolve the conflicting file before testing :)

**Comment by zuk3975** [comment]
@sarahdib rebased. The navbar disappeared :smile: (`It is not related to the list though, it came from develop branch` #15309), so you can still test the list using link provided in description.

**Review by matks — DISMISSED** [review]

**Comment by matks** [comment]
PrettyCI ✅ 
Review ✅

**Comment by matks** [comment]
@sarahdib Can you perform the QA validation and then we'll rebase ?

Please do not ask us to rebase again 🙏everytime we do it, 24 hours later we need a new rebase 😅

**Comment by zuk3975** [comment]
> @sarahdib Can you perform the QA validation and then we'll rebase ?
> 
> Please do not ask us to rebase again everytime we do it, 24 hours later we need a new rebase

yes, please, this is very small conflict :smile:

**Comment by sarahdib** [comment]
@matks & @zuk3975 I understand your point but in QA we need to have a PR without conflicting files.

**Comment by matks** [comment]
> @matks & @zuk3975 I understand your point but in QA we need to have a PR without conflicting files.

But it's only the assets 😉 

Can we re-use a way of collaboration we had setup with @marionf ? Because the assets are files that very often re-computed totally, there is a lot (I mean a HUUUUGE lot) of times where assets are the files being in conflict. So what we were doing with marion is that we were stating in the PR description "please rebuild assets" and she would rebuild them in order to have fresh assets install.

Assets can be outdated but never wrong because they are fully computed. When we rebase we just recompute them.

Would this work @sarahdib ?

**Comment by sarahdib** [comment]
@matks Ok I see I will  test it :)

**Comment by sarahdib** [comment]
Hello @zuk3975 

It's ok for me.

Just the 3rd problem with calendar in the little screen. I let @TristanLDD respond about that 
61435383-a13e7b00-a938-11e9-9ff8-fce8dbf4afd1

Thank you

**Comment by zuk3975** [comment]
> Hello @zuk3975
> 
> It's ok for me.
> 
> Just the 3rd problem with calendar in the little screen. I let @TristanLDD respond about that
> 61435383-a13e7b00-a938-11e9-9ff8-fce8dbf4afd1
> 
> Thank you

hmm. I think it is global grid problem for lists with many fields. Guess it should be fixed in another PR. @matks, @sarjon wdyt?

**Comment by sarjon** [comment]
Yeah, seems this problem is not related to this PR, but to all grids instead. @matks will you pass this?

**Comment by matks** [comment]
Created dedicated issue https://github.com/PrestaShop/PrestaShop/issues/15534 for that

@sarahdib can we mark it as `QA ✔️" then ?

**Comment by matks** [comment]
@zuk3975 One rebase and we can merge ! ❤️

**Comment by zuk3975** [comment]
@matks, rebased :smile:

**Comment by matks** [comment]
Thank you @zuk3975 ! 🎉

**Review by matks — APPROVED** [review]

File: src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php
Comment (matks) [reply src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php:179]: Fixed in https://github.com/PrestaShop/PrestaShop/pull/18691 😄  almost one year later

**Pull request merged**
